design/baremetal-operator/sub-states.md

```diff
@@ -0,0 +1,155 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost sub-states
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes a new `BareMetalHost` field `SubState` to specify more
+precisely at which point in a BMH lifecycle it is.
+
+## Motivation
+
+The primary motivation is to enable operators an easier insight into what
+is going on with their nodes. This simplifies debugging and makes people feel
+better about long-running processes. It may also be helpful for UI tools
+to be able to display some sort of progress information.
+
+### Goals
+
+* Provide a way to know what action is being executed right now without
+  accessing Ironic API.
+
+### Non-Goals
+
+* Provide an access to the Ironic Node `provision_state` field.
+* Establish a direct one-to-one mapping between provision states and the
+  new field values.
+* Provide more granularity than what the Ironic API can report.
+* The values of the new field will not be a part of the API contract and can
+  change as a result of backend changes.
+* Define a non-empty value for each possible situation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator reasonably familiar with the provisioning process, I would like
+to know what action is being executed on a BMH right now.
+
+## Design Details
+
+<!-- dtantsur: I'm terribly unfamiliar with the code, help appreciated -->
+
+* New field `SubState` is added to `BareMetalHost`
+* The value of the field is re-calculated based on the result described below
+  on each reconciliation.
+
+### Implementation Details/Notes/Constraints
+
+The new field will be populated based on the following rules, where `Node` is
+an Ironic Node object. The rules are defined in pseudo-code.
+
+#### For "Provisioning" state
+
+* IF `Node.provision_state` in "cleaning", "clean wait":
+
+  * IF `Node.clean_step` is null:
+
+    SubState is `Preparing Cleaning`.
+
+  * IF `Node.clean_step["step"]` in "erase\_devices",
+    "erase\_devices\_metadata":
+
+    SubState is `Cleaning Disks`.
+
+* IF `Node.provision_state` in "deploying", "deploy call-back":
+
+  * IF `Node.deploy_step` is null OR `Node.deploy_step["step"]` is "deploy":
+
+    SubState is `Preparing Deployment`.
+
+  * IF `Node.deploy_step["step"]` is "write\_image":
+
+    SubState is `Writing Image`.
+
+  * ELSE:
+
+    SubState is `Configuring`.
+
+<!-- dtantsur: it's possible to be more granular, but it may not be justified,
+especially since only the write_image step runs noticable amount of time -->
+
+#### For the planned "Preparing" state
+
+This relies on [the RAID
+design](https://github.com/metal3-io/metal3-docs/pull/134) accepted.
+
+* IF `Node.provision_state` in "cleaning", "clean wait":
+
+  * IF `Node.clean_step` is null:
+
+    SubState is `Preparing Configuration`.
```

Comment:
Is this about preparing to configure the RAID controller, then?

Reply:
It's the same as "Preparing cleaning" above, but I'm trying to avoid using "cleaning" to mean anything except for "cleaning disks". This includes booting the agent (if without fast-track) and doing other small preparation items.